### DISCUSSION TITLE
2023 Planning

### DISCUSSION THREAD

**Discussion by zackkrida** [WordPress/openverse#343] [Ideas]
This discussion is a conversation surrounding 2023 plans for the Openverse project. There have been several disparate discussions in different channels and formats. This discussion aims to aggregate those conversations into one place and come up with a process for prioritizing and discussing Openverse goals and projects for the upcoming year.

## Due Date

2022-12-22

## Prior discussions

- [Monthly Priorities Meeting Notes](https://docs.google.com/document/d/1aUhpakIUy4we0BU0D3Pc7hqMJg_FpyxoB6eulBw6Z9c/edit#)
- [Thinking towards 2023 Post](https://make.wordpress.org/openverse/2022/11/16/thinking-towards-2023/)
- [Preparing the next migration of the Catalog](https://make.wordpress.org/openverse/2022/11/18/preparing-the-next-migration-of-the-catalog/)

**Comment by zackkrida** [comment]
## Project Prioritization

### Prerequisite Todos

- [ ] Add proposed projects to the sheet [WIP]

### Assignment

For each project in the [project sheet](https://docs.google.com/spreadsheets/d/1sj9dgkJIBhJ5bYfx5iogWEPjJQDw6d6rQKtnTHlYLbg/edit?usp=sharing)

- Link it to a corresponding lighthouse goal
- Identify possible participants and their roles
- Rank the project on a scale of 1-5 for the scope, with 1 being “smallest” and 5 being “largest”.

Any issues or concerns with this process can be discussed here.

**Reply by sarayourfriend** [reply]
Something that came up briefly when Staci and I were talking was about the "content safety" goals. Specifically, that it felt like we would maybe want to do some up-front planning to identify the potential projects that would fall under a general content safety goal. Basically, a separate planning process for each of the overarching goals to explore the kinds of projects that could fit into it and give brief summaries. Just for content safety, for example, there are at least these things in the spreadsheet: blurring images on single result pages, increasing inputs for determining whether a result is "mature/NSFW" (this has two distinct sub-projects and some long-ranging effects as far as how we use the data from the different inputs), developing a process for triaging mature media reports, making Sentry errors safer by filtering keywords. Are there other things? I just want to make sure that each of these is captured so that we can estimate them and prioritise them accordingly without leaving out some gaps.

As folks work on those, by the way, it would be good to try to consider any assumptions or prerequisites that exist for any of the individual projects that might themselves be projects.

Would taking a moment and assigning a lighthouse goal to individuals to explore and brainstorm project ideas be worthwhile as a step in this process? Is it just enough to put everything we can think of into the project list? Additionally, how big is a "project" is another question I am trying to wrap my head around. Is switching to use UPDATE_INDEX a "project" or a one-off issue that doesn't require much planning? Hard to tell, so maybe it's best to expect it would require at least some brief project planning? Worst case we have a well-researched issue when we're ready to tackle it.

Additionally, what are the actual "lighthouse goals"? Several were thrown out during the Make discussion. It seems like potentially a list of specific ones was settled on at some point, but I'm not sure what list to refer to here when trying to suggest the lighthouse goal for a given project. Something I'm unsure of, for example, is whether "data inertia" is a lighthouse goal or a collection of projects? It feels about as abstract and large as "content safety" but it wasn't something that was discussed in the meeting notes (unless I missed it).

**Reply by sarayourfriend** [reply]
A couple of more questions, sorry! When considering the "scale" of a project, are we considering the expertise of the people who've potentially assigned themselves to the project? I ask because there are some projects which would naturally take longer or shorter for different people on the team to work on due to their relevant experiences. Additionally, what happens if we scope the effort and then plan a project based on subject-matter experts being involved, but then those people stop contributing to Openverse? This latter question leads me to think we would want to estimate effort based on an abstract, naive contributor who only has a basic general understanding of any given part of Openverse. That way we might better achieve a scale estimation that would more closely even out to "anyone working on any project" rather than SMEs.

Does that make sense to other folks or is there a different mindset that would be more productive?

Also: what is the best way to contribute to the "scale" column? Should the person who adds the project estimate the scale? Do we want to try voting on the scale using one of those tools that only shows the votes after everyone has voted so that people don't accidentally influence each other and under or overestimate based on particular assumptions? I don't know if these tools work "asynchronously" so that everyone could participate over a given 24-hour period or something like that.

**Reply by zackkrida** [reply]
Great questions, thanks @sarayourfriend. 

## On content safety

To make sure I'm understanding correctly, it sounds like you're suggesting we do some research and planning where we attempt to outline our general goals and reasoning behind content safety work, as a means of making sure the current projects are a good fit but also that we're not missing other potential projects?

I think that's a good idea. This is such a far-reaching and big subject. For example, I haven't seen much discussion yet about we have to contend with "mature" content which can be shown in Openverse, but also "mature" content that we don't want to index at all or want to report for illegality. This planning would also be a good opportunity to identify potential partners for the sub-projects. 

Perhaps we could treat this "Content Safety Planning" as a standalone project itself, the first we do for content safety, so we can follow the RFC and implementation processes? In this case, the planning RFC would identify potential projects, collaborators, goals, and motivations, and the implementation plan would identify an order in which to tackle the projects. Does that sound like a solid approach? Any other ideas?

## On this "2023 Planning" process

> Is it just enough to put everything we can think of into the project list?

I _think so_, primarily because I think we already have more items here than fit into a single year. At this stage I think we have sufficient ideas and the prioritization is more important. I want it to be easy for folks to add things to the sheet and I think we've already spent a fair bit of time and a few sync sessions identifying projects. 

If anyone thinks this would be useful, however, feel free to set something up for next week!

> Would taking a moment and assigning a lighthouse goal to individuals to explore and brainstorm project ideas be worthwhile as a step in this process?

I think for future efforts this would be a really great idea. Especially as we move away from a mentality where individuals work per-repository/slice of the stack, and towards one more focused on projects, I could see a future where we split into squads yearly, with a squad focused on one goal/aspect of the project (relevancy squad, safety squad, etc.). We'll need to figure out what balance of specificity vs. the ability to jump around folks value most.

To reiterate at this stage I think we can focus on listing and prioritizing the projects we _already have_ in the notes and from past discussions.

> Additionally, what are the actual "lighthouse goals"?

I should have created it before the project list, I wanted a top-level comment here for confirming the list of goals. I'll do so now. 

> Something I'm unsure of, for example, is whether "data inertia" is a lighthouse goal or a collection of projects?

My inclination would be the later. I think fixing "data inertia" should _be in service of_ other goals, not something to do by it's own virtue.

### Project Scale

> When considering the "scale" of a project, are we considering the expertise of the people who've potentially assigned themselves to the project?

I think we should try to do this from the perspective of 'any contributor', and factor in:

- The time to produce documentation, based on the complexity of the project at hand
- The time to implement the actual work, provided that the issues have been clearly documented so that any Openverse contributor, sponsored or a subject matter or otherwise, can execute the work.

> What is the best way to contribute to the "scale" column?

I'm happy to adhere to any system, my default assumption, which should have been made explicit, apologies, was that anyone could fill it out for any column based on their own thinking, and then folks could comment if they disagree, propose alternative numbers, etc. If people want something more structured I'm happy to help set that up.

## On project size

Still refining my own ideas around this, but one quick thought is that we can find ways to determine if a work stream is project-worthy or not based on its size, duration, and how much tracking it needs. For things that we don't track as a 'project thread' (a process we hope to implement in GitHub moving forward), we might still track them in other ways:

1. Dated Milestones for grouped/related issues or multi-issue features
2. We may still write RFCs for large, one-off issues that don't need project tracking
3. Using a date column on a GitHub project
4. A dedicated 'project thread' for small, non-standalone projects

**Reply by sarayourfriend** [reply]
# Content Safety

> To make sure I'm understanding correctly, it sounds like you're suggesting we do some research and planning where we attempt to outline our general goals and reasoning behind content safety work, as a means of making sure the current projects are a good fit but also that we're not missing other potential projects?

For content safety, this was a specific recommendation that @stacimc made earlier. What I was trying to get at was whether this was something we wanted to apply to other projects as well.

As for the specific idea for content safety that you shared, I think it is a good one. However, there are things that we _know_ need to happen for content safety. For example, we know that we need to create a process of triaging and addressing content reports of all kinds but especially mature content reports. This will require non-trivial API and infrastructure work to make it fully possible as well as collaboration with existing WP.org moderation teams (like the WP Photo Directory) team. Likewise, getting more and quality data about mature image classification is something we know will benefit us regardless, so we can confidently schedule that if we think it is something we can fit into the year.

Basically: I am confident that we can identify the broad strokes of a lighthouse goal and some of the most important aspects of it. However, for some of them, especially content safety, where I don't think anyone on the team is particular expert, it is probably important for us to spend specific time learning about the issue and consulting with experts on it. Given the Openverse team has excellent access to several different content moderation professionals, it would be negligent not to consult with them about it. Particularly because the consequences of getting the process wrong can be dire, as can delaying it.

# Planning process

> At this stage I think we have sufficient ideas and the prioritization is more important

We have a lot of project ideas but I don't know if they're equally spread across the lighthouse goals. I added a lot of relevancy and content safety ones, but did not add many data inertia ones. It would be great if @WordPress/openverse-catalog folks could add their ideas for data specific projects. Likewise for folks working closely on other parts of the project that are not currently well represented on the spreadsheet.

> I think for future efforts this would be a really great idea. Especially as we move away from a mentality where individuals work per-repository/slice of the stack, and towards one more focused on projects, I could see a future where we split into squads yearly, with a squad focused on one goal/aspect of the project (relevancy squad, safety squad, etc.). We'll need to figure out what balance of specificity vs. the ability to jump around folks value most.

Sounds cool!

> My inclination would be the later. I think fixing "data inertia" should be in service of other goals, not something to do by it's own virtue.

Gotcha. I'll leave it to the catalogue folks to discuss this one. I'll ping @aetherunbound @stacimc and @rmwidom specifically for their thoughts here.

My own impression is that data inertia is such a significant set of projects (data infrastructure, refresh redesigns, etc) that it could use it's own lighthouse goal and series of distinct projects. If it starts to take longer than a week to refresh the API database then we'll end up with a critical backlog of data that never ends up available in Openverse API... which seems fundamentally antithetical to our goals. To me it seems self-evident that it is important to work on for its own merits. Perhaps there are other ways of justifying this work that you had in mind, though, Zack?

> I'm happy to adhere to any system [for scale estimation], my default assumption, which should have been made explicit, apologies, was that anyone could fill it out for any column based on their own thinking, and then folks could comment if they disagree, propose alternative numbers, etc. If people want something more structured I'm happy to help set that up.

My fear with this is that it is easy for someone to accidentally dominate scale estimation on a large number of projects.

Additionally, the scale currently feels arbitrary to me. Could we identify an example of a 1 and an example of a 5? For me, at least, it would be helpful for conceptualising what would fit into the other 3 scale estimates.

Thanks for the detailed responses to my questions!

**Reply by AetherUnbound** [reply]
### _Regarding data inertia_

I actually do think having that as a lighthouse goal (or maybe something _slightly_ broader) would be a good idea! As Sara mentions, there's a lot that factors into how massive and unwieldy our data is. The `UPDATE_INDEX` data refresh change is one piece, but also the data cleaning (https://github.com/WordPress/openverse/pull/345) and database migration work that's being done as part of the Data Normalization process. With us migrating the API to ECS, there's also the data inertia we have to deal with regarding database migrations there too. Perhaps the lighthouse goal could be broader to include data health and quality as well, but it seems that we have at least several potential projects under that umbrella already.

**Reply by stacimc** [reply]
## Data Inertia

+1 from me as well for having an additional Data lighthouse goal. Data Inertia, to me, encapsulates:
* stabilizing/speeding up the data refresh
* improving data migrations
* literally adding new providers (and reducing the difficulty of adding new providers, for example through the base class)
* larger, fundamental changes to the pipeline

All with the goal of getting accurate data from source providers into the API as quickly as possible. Arguably many data health/quality projects could already be considered part of this lighthouse goal, IMO.

**Reply by fcoveram** [reply]
## On this "2023 Planning" process
### Project size

My main take on the discussion is trying to keep the process as simple as possible to avoid future frustrations with unaccomplished expectations. It sounds simplistic, but the need that drove us to address project documentation was solidify the project goal, scope, stakeholders, status, and tasks in a single place.

Regarding when starting a project is appropriate, I would rely on folks' judgment as long as we have a set of principles or similar to _start_ the initiative. Take into account that weekly and prioritization meetings allow us to revisit the actions taken.

### Project scale

From previous experiences working under Lean UX and Agile methodologies, scale assignation mainly works to detect discrepancies among evaluations. Participants discuss why a project was assigned with 1 and 5 at the same time to conclude specific actions. So I would not overwhelm with personal scorings until having an outcome to revisit.

## On content safety

> Content safety, where I don't think anyone on the team is particular expert, it is probably important for us to spend specific time learning about the issue and consulting with experts on it.

I suggest covering this topic in the [2023 WordPress Community Summit](https://make.wordpress.org/community/2022/12/08/suggest-topics-for-the-2023-wordpress-community-summit/) proposal. Involving the community beforehand would help us to include unknown-unknowns and address future actions.

**Reply by sarayourfriend** [reply]
I don't know where we landed on scoring. It seems like there are differences in opinion in how to do this (my own opinion, borne from experience, matches Francisco's). But there's only a week left for discussion and Zack is AFK, so there isn't anyone to give us a definitive answer. I checked the spreadsheet and folks had already added effort scorings, so I added some to other things because I am anxious about this conversation not resolving before the start of the year, which was the original goal. I did this to hopefully move the conversation along so that we can call it closed at some point.

If we want to do a different way that does not have a small group of people having an outsized influence on the perceived effort of a project and there is buy in from @WordPress/openverse-maintainers (please do respond here about what you want to be done in this regard, as soon as you're able to during your work week if you're online), then we can find a different tool to do it and accomplish the goal that Francisco mentioned of identifying discrepancies.

**Reply by AetherUnbound** [reply]
@sarayourfriend I am happy to try out a tool as you and @panchovm describe for scale estimation which (as I understand it) would have us all provide estimations and allow us the capacity to see discrepancies. I do not have experience with this technique (or granular estimation in general, really) so I don't have any strong opinions against such an approach.

**Reply by sarayourfriend** [reply]
I can't find any free tools for doing it asynchronously so that people aren't influenced by previous voters, which is crucial for uncovering discrepancies. Wikipedia has a good explanation of the idea and reasons for doing it this way: https://en.wikipedia.org/wiki/Planning_poker#Rationale

The only easy alternative I can think of is for each of us to copy the list of projects from the spreadsheet into a personal sheet in the project scratchpad linked. Then, once everyone who wants to participate has participated (hopefully this would be as soon as possible because we're short on time), we can identify the major discrepancies and discuss them.

@WordPress/openverse-maintainers It would be great to hear from additional folks about this unless this project planning isn't a priority, and I'm feeling an undue sense of urgency to resolve the discussion. If that is the case, then please let me know and I will stop pushing for this to get done and annoying folks with pings.

**Reply by AetherUnbound** [reply]
For what it's worth I personally don't feel a sense of urgency in resolving this aspect of the project planning before the deadline. Perhaps my own priorities are misaligned and I'm not sure if that helps, but I think we'll certainly want to assign scale before we start deciding which of these projects to focus on (which seems doubtful to happen in the next two weeks).

**Reply by fcoveram** [reply]
It seems that we agree with each member scoring ideas and then discussing discrepancies. I have thinking about this, and FigJam coulb be a good tool for the following steps.

#### One project per post-it
Create a FigJam file, add the Lighthouse goals as sections, and move each idea into its associated goal.

#### Scoring each project
Each of us scores all ideas on a 1-to-5 scale. The score comes from our personal backgrounds, so it is ok to miss technical knowledge. The score is for detecting discrepancies and discussing them later.

#### Discuss discrepancies
Since most of us will attend an IRL meeting, we can either discuss the discrepancies there and add the final score in the FigJam file, or discuss them async here in this post, and add the final score once we reach a certain consent.

What I am missing from the above is the _prioritization list for the forthcoming tasks_. The funnel we are following is:
1. Defining the 2023 goals → Done ✅ 
2. Brainstorming projects for each goal → Done ✅ 
3. Scoring those projects from the effort side → In progress 🎯 
4. Solving discrepancies in the efforts required.

But then I miss the prioritization stage, where we list what comes first and spread them out through milestones within a tentative 2023 time schedule.

The effort scoring is great for assessing our tasks, but it does not take into account *how impactful they are*. For instance, the “Perceptual image quality detection” project requires higher efforts than “Browsing through shortcuts”, as noted in the spreadsheet, but it might impact less how visitors use Openverse.

For this step, I propose adding two more steps to the list.

#### 5.Impactful scoring
On the same scale of 1 to 5, score each project on how impactful for users we think it is. Again, a non-technical approach aiming to detect discrepancies

#### 6.Prioritization matrix
Based on impact and effort scores, move each project into an “Impact vs Effort” matrix to have a clearer picture of projects and decide next steps.

---

What do you think of this idea? Indeed I can create the FigJam file and move all the work made.

**Reply by AetherUnbound** [reply]
This is an excellent idea Francisco, thank you for outlining it so clearly!

**Reply by sarayourfriend** [reply]
I personally find FigJam really difficult to work with and even harder to get an idea of what went on after the fact of a meeting. But if it is the preference of most of the team then I of course will participate.

My only concern with the plan is that it does not account for having scores not be revealed (or at least nominally hidden) until everyone who wants to has voted on a score. My understanding of scoring is that it's important for scores to be revealed all at the same time, otherwise people's scores end up being implicitly influenced by the people who've shared their scores before them.

I do really like having the impact score as well.

As for the prioritisation matrix, I previously suggested in the Google document notes for the initial meeting from a couple weeks ago that we use a Gantt chart for that and there are several very nice tools for building Gantt charts. It would allow us to plot projects on a timeline all together in a linear, easy to understand view and easily reveal dependencies.

**Reply by zackkrida** [reply]
@WordPress/openverse I have created a template for individually ranking projects according to scale and impact. Please see the template for full instructions:

https://docs.google.com/spreadsheets/d/1OXBLGdsFisiUh6o73rTLo_DOAmunQCaDGC1UzCZ3bn0/edit#gid=0

If all the sponsored contributors could aim to complete this by **2023-01-13**, that would be ideal. Thank you!

**Comment by zackkrida** [comment]
## Lighthouse Goals

_"Lighthouse" goals refers to goals which inform and illuminate a path forward for Openverse contributors. These are concrete things to be accomplished but also principles that can informa and steer all of our work for the coming year._

Let's discuss and refine what our lighthouse goals for 2023 should be. To resurface the ideas already shared that I've seen consensus around:

- Quantify our work. 
- Spread awareness of Openverse.
- Improve search result relevancy. 
- Reduce data inertia.
- Content safety.

Additional ideas that either weren't discussed in detail to my knowledge:

- Improving user search experience on the front end: make it easier to search by provider/tag/creator.
- Devex/Tech debt [Things like the Nuxt 3 migration]
 
In the comments here we can attempt to settle on a final list.

**Reply by zackkrida** [reply]
I'm thinking a lot about the 'spread awareness' goal and would like to find a way for it to encompass more things. To not just be about advocacy/evangelism, but to capture more things like:

- Community contributors and the quality of their experience
- Outreach about Openverse at WordCamps and other conferences
- Partnerships with existing image providers, potential providers
- Partnerships with broader 'open access' movements and organizations

Perhaps something about community would be better phrasing. I'll keep thinking about it.

**Reply by sarayourfriend** [reply]
> Devex/Tech debt [Things like the Nuxt 3 migration]

One idea for addressing this could be to allocate 10% of our time for the next year (or some other percentage) to specific DevEx improvements, especially gnarly ones like Nuxt upgrade, that we know need to happen.

> Improving user search experience on the front end: make it easier to search by provider/tag/creator.

Would these fall under relevancy, or are we _only_ speaking about changes to the API results when we discuss relevancy? To me, anything that helps the user define to us what would be more relevant results is also a win for overall relevancy. I may be thinking _too_ expansively about relevancy though, in which case it would be helpful if we re-worded the search relevancy goal.

**Reply by sarayourfriend** [reply]
> I'm thinking a lot about the 'spread awareness' goal and would like to find a way for it to encompass more things. To not just be about advocacy/evangelism, but to capture more things like:

Could it be: "Make Openverse easier to contribute, understand, and collaborate on"? This would encompass the things you mentioned as not necessarily being captured from simply "spread awareness".

**Reply by stacimc** [reply]
> > Devex/Tech debt [Things like the Nuxt 3 migration]
>
> One idea for addressing this could be to allocate 10% of our time for the next year (or some other percentage) to specific DevEx improvements, especially gnarly ones like Nuxt upgrade, that we know need to happen.

I would love to have an always-open DevEx stream of work, ideally collected into its own board/thread so it can be constantly added to as new issues are created, and periodically re-prioritized. I think the 10% idea would work especially well if defined as "10% of the **team's** time" rather than a flat 10% across all team members -- so, at the beginning of a two-week sprint we could identify a couple of issues from the DevEx stream of work and assign them.

**Reply by obulat** [reply]
> > Improving user search experience on the front end: make it easier to search by provider/tag/creator.
> 
> Would these fall under relevancy, or are we _only_ speaking about changes to the API results when we discuss relevancy? To me, anything that helps the user define to us what would be more relevant results is also a win for overall relevancy. I may be thinking _too_ expansively about relevancy though, in which case it would be helpful if we re-worded the search relevancy goal.

I agree with @sarayourfriend , the lighthouse goal of **improving the search experience** could comprise the two parts
- search relevancy improvements in the API. This would affect all users, from the front end to API to WordPress.com/Gutenberg. Changes in the catalog that add more data (e.g. image quality, tags from Recognition or descriptions that are subsequently used to improve the Elastic Search results) can also be included here.
- the improvements in the search experience for the users of the front end specifically: additional ways of filtering and searching, pages for providers, better visibility for filters.

**Reply by fcoveram** [reply]
## Spread awareness of Openverse

Very much agree with @zackkrida on the subsequent points.

## Improving user search experience on the front end

I agree with adding the "improving user search experience on the front end" point as a lighthouse goal, but the "search by provider/tag/creator" is a specific task that I am not sure impacts bigger than other pending improvements like the modal browsing. Going back and forth between single results still feels forceful as it requires clicking on "back to results." 

This does not block us from working on features to see filtered results from the single result page, but my prioritization goes to the modal implementation first.

**Reply by sarayourfriend** [reply]
> This does not block us from working on features to see filtered results from the single result page, but my prioritization goes to the modal implementation first.

I added the modal results to the project spreadsheet :+1: Which lighthouse goal do you think it falls under, @panchovm? I feel like it is actually leftover work from the redesign goal we had last year...

**Reply by fcoveram** [reply]
Oh, I completely forgot to mention that @sarayourfriend! I think it belongs to a goal _by itself_. Pondering that these goals will generate a set of tasks, "search by provider/tag/creator" is one of them, alongside other similars like "sorting results."

**Reply by obulat** [reply]
> > This does not block us from working on features to see filtered results from the single result page, but my prioritization goes to the modal implementation first.
> 
> I added the modal results to the project spreadsheet :+1: Which lighthouse goal do you think it falls under, @panchovm? I feel like it is actually leftover work from the redesign goal we had last year...

I see this as part of the  "Improving search experience" (the frontend side). 
Other frontend tasks can improve the _perceived_ relevancy because the search results can become more relevant when you select providers, or tags, or sort by update time.
Adding the modal can improve the user flow for selecting the most relevant result. I don't think that it needs a full-blown project planning process, but it might be nice to collect metrics before/after.

**Reply by obulat** [reply]
> I'm thinking a lot about the 'spread awareness' goal and would like to find a way for it to encompass more things. To not just be about advocacy/evangelism, but to capture more things like:
> 
> - Community contributors and the quality of their experience
> - Outreach about Openverse at WordCamps and other conferences
> - Partnerships with existing image providers, potential providers
> - Partnerships with broader 'open access' movements and organizations
> 
> Perhaps something about community would be better phrasing. I'll keep thinking about it.

Maybe something like "Develop Openverse community" could be a good name?

**Comment by sarayourfriend** [comment]
Once we have a specific set of lighthouse goals for 2023, could we create issue labels for them to apply to individual issues? That could help us strike a balance in driving towards each goal and ensure that none of them fall by the wayside due to low visibility or getting forgotten about in the midst of other exciting work.

**Reply by AetherUnbound** [reply]
That's a great idea!

**Reply by zackkrida** [reply]
This would be very nice in a dashboard / GitHub project view as well with a column or section for each goal, which could give us a sense of which goals we're allocating the most time to.
